components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal

```diff
@@ -0,0 +1,157 @@
+
+    public function filterRequest(http:Caller caller, http:Request request, http:FilterContext context)
+                        returns boolean {
+        //Setting UUID
+        int startingTime = getCurrentTime();
+        context.attributes[REQUEST_TIME] = startingTime;
+        checkOrSetMessageID(context);
+        setHostHeaderToFilterContext(request, context);
+        setLatency(startingTime, context, SECURITY_LATENCY_AUTHN);
+        return doAuthnFilterRequest(caller, request, context);
+        
+    }
+
+    public function filterResponse(http:Response response, http:FilterContext context) returns boolean {
+        return true;
+    }
+};
+
+function doAuthnFilterRequest(http:Caller caller, http:Request request, http:FilterContext context)
+             returns boolean {
+    boolean isOauth2Enabled = false;
+    runtime:InvocationContext invocationContext = runtime:getInvocationContext();
+    invocationContext.attributes[MESSAGE_ID] = <string>context.attributes[MESSAGE_ID];
+    printDebug(KEY_AUTHN_FILTER, "Processing request via Pre Authentication filter.");
+
+    context.attributes[REMOTE_ADDRESS] = getClientIp(request, caller);
+    context.attributes[FILTER_FAILED] = false;
+    invocationContext.attributes[SERVICE_TYPE_ATTR] = context.getService();
+    invocationContext.attributes[RESOURCE_NAME_ATTR] = context.getResourceName();
+
+    boolean isCookie = false;
+    string authHeader = "";
+    string? authCookie = "";
+    string|error extractedToken = "";
+    string authHeaderName = getAuthorizationHeader(invocationContext);
+    invocationContext.attributes[AUTH_HEADER] = authHeaderName;
+    string[] authProvidersIds = getAuthProviders(context.getServiceName());
+
+    if (request.hasHeader(authHeaderName)) {
+            authHeader = request.getHeader(authHeaderName);
+    } else if (request.hasHeader(COOKIE_HEADER)){
+            //Authentiction with HTTP cookies
+        isCookie = config:contains(COOKIE_HEADER);
+        if (isCookie) {
+            authCookie = getAuthCookieIfPresent(request);
+            if (authCookie is string) {
+                authHeader = authCookie;
+            }
+        }
+    }
+    
+    string providerId;
+    if (!isCookie) {
+        providerId = getAuthenticationProviderType(authHeader);
+    } else {
+        providerId = getAuthenticationProviderTypeWithCookie(authHeader);
+    }
+    boolean canHandleAuthentication = false;
+    foreach string provider in authProvidersIds {
+        if(provider == providerId) {
+            canHandleAuthentication = true;
+        }
+    }
+
+    if(!canHandleAuthentication) {
+        setErrorMessageToFilterContext(context, API_AUTH_PROVIDER_INVALID);
+        sendErrorResponse(caller, request, context);
+        return false;
+    }
+    // if auth providers are there, use those to authenticate
+    
+        //TODO: Move this to post authentication handler
+        //checkAndRemoveAuthHeaders(request, authHeaderName);
```

Comment:
Is this `TODO:` still something we need to work on or resolved?

Reply:
We need to work on this, not resolved